scheduler/queue/concurrent_random_queue.go

```diff
@@ -0,0 +1,40 @@
+package queue
+
+import (
+	"math/rand"
+	"sync"
+)
+
+// ConcurrentRandomQueue is a generic, thread-safe queue
+// that pops random elements in O(1) time.
+type ConcurrentRandomQueue[T any] struct {
+	mu    sync.Mutex
+	queue []T
+}
+
+func NewConcurrentRandomQueue[T any](capacityHint int) *ConcurrentRandomQueue[T] {
+	return &ConcurrentRandomQueue[T]{queue: make([]T, 0, capacityHint)}
+}
+
+func (q *ConcurrentRandomQueue[T]) Push(item T) {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+
+	q.queue = append(q.queue, item)
+}
+
+func (q *ConcurrentRandomQueue[T]) Pop() *T {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+
+	if len(q.queue) == 0 {
+		return nil
+	}
+	idx := rand.Intn(len(q.queue))
```

Comment:
I like the randomness idea 💯 

Maybe we can use a seed like we do here https://github.com/cloudquery/plugin-sdk/blob/00b9d9a6c93bad24b0ea7d489ef13d65cead2023/scheduler/scheduler_shuffle.go#L46

so the order is somewhat the same if your table list doesn't change (e.g. you re-run the same sync)

Reply:
I passed the seed to the Scheduler's constructor, and from it to the Random Queue.